### PR TITLE
Add parseUrl function for Bitbucket

### PR DESCRIPTION
**Pull request by d2lam**
- Takes care of https and ssh url
- We need to pass these fields to make calls to Bitbucket API:
```
login_type: 'oauth2',
oauth_access_token: config.token
```

Relates to https://github.com/screwdriver-cd/screwdriver/issues/255

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,18 @@
...
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| config             | Object | Configuration Object |
+| config.checkoutUrl | String | Checkout url for a repo to parse |
+| config.token  | String | Access token for scm |
+
+#### Output: Promise
+1. Resolves to an scm uri for the repository. Ex: `github.com:1234:branchName`, where 1234 is repo ID.
```
Comment (stjohnjohnson) [inline README.md]: bitbucket.com?

File: index.js [hunk 2]
```diff
@@ -3,8 +3,73 @@
 const Scm = require('screwdriver-scm-base');
 const hoek = require('hoek');
 const url = require('url');
+const request = require('request');
+const schema = require('screwdriver-data-schema');
+const API_URL = 'https://api.bitbucket.org/2.0/repositories/';
```
Comment (stjohnjohnson) [inline index.js]: This is hardcoded for bitbucket.org, we should allow users to override this on setup

File: index.js [hunk 3]
```diff
@@ -3,8 +3,73 @@
...
+    const regex = schema.config.regex.CHECKOUT_URL;
+    const matched = regex.exec(checkoutUrl);
+    const result = {
+        username: matched[MATCH_COMPONENT_USER],
+        repo: matched[MATCH_COMPONENT_REPO],
+        branch: matched[MATCH_COMPONENT_BRANCH].slice(1)
+    };
+    let host;
+
+    if (checkoutUrl.indexOf('https') > -1) {
```
Comment (stjohnjohnson) [inline index.js]: What is this for?  If you are checking for http or https, it should be checking for `=== 0`

Comment (stjohnjohnson) [inline index.js]: Better Regex:
```
/^(?:(?:https?|git):\/\/)?(?:[^@]+@)?([^/:]+)(?:/|:)([^\/]+)\/([^.]+)(?:\.git)?(#.+)?$/
```

Matches:
 - https://stjohnjohnson@bitbucket.org/stjohnjohnson/foobar
 - git@bitbucket.org:stjohnjohnson/foobar.git
 - git@github.com:screwdriver-cd/data-schema.git
 - https://github.com/screwdriver-cd/data-schema.git

With result:
 - 0 is hostname
 - 1 is org
 - 2 is repo
 - 3 is branch

File: index.js [hunk 4]
```diff
@@ -3,8 +3,73 @@
...
+            `${API_URL}${repoInfo.username}/${repoInfo.repo}/refs/branches/${repoInfo.branch}`;
+        const options = {
+            url: getBranchUrl,
+            method: 'GET',
+            login_type: 'oauth2',
+            oauth_access_token: config.token
+        };
+
+        return new Promise((resolve, reject) =>
+            request(options, (err, response, body) => {
```
Comment (stjohnjohnson) [inline index.js]: What about the breaker?

File: test/index.test.js [hunk 5]
```diff
@@ -12,15 +12,98 @@ sinon.assert.expose(assert, { prefix: '' });
...
+            };
+
+            requestMock.yieldsAsync(null, {}, body);
+
+            const expected = 'bitbucket.org:{de7d7695-1196-46a1-b87d-371b7b2945ab}:master';
+
+            return scm.parseUrl({
+                checkoutUrl: 'git@bitbucket.org:batman/test.git#master',
+                token: 'myAccessToken'
+            }).then(parsed => assert.equal(parsed, expected));
```
Comment (tkyi) [inline test/index.test.js]: Should we add some `assert.callWith` tests for `request`?

File: index.js (hunk 3 above)
Comment (d2lam) [inline index.js]: Cool thanks! I was doing it off `schema.config.regex.CHECKOUT_URL`
We should probably change that to this regex.

File: index.js (hunk 2 above)
Comment (d2lam) [inline index.js]: This will be done after we have the `setup()` right? Or you want me to add that now too?

File: index.js (hunk 4 above)
Comment (d2lam) [inline index.js]: Will add it

File: index.js (hunk 3 above)
Comment (tkyi) [inline index.js]: Making a pull request to change that

File: index.js (hunk 2 above)
Comment (stjohnjohnson) [inline index.js]: We can wait, just wanted to let you know.

File: test/index.test.js (hunk 5 above)
Comment (d2lam) [inline test/index.test.js]: Added, but changed to breaker

File: index.js (hunk 4 above)
Comment (d2lam) [inline index.js]: Added

File: README.md (hunk 1 above)
Comment (d2lam) [inline README.md]: Changed

File: index.js (hunk 3 above)
Comment (d2lam) [inline index.js]: Waiting to use it

**Comment by d2lam** [comment]
Removed implementation for stats and just calls `this.breaker.stats()` thanks to https://github.com/screwdriver-cd/circuit-fuses/pull/9 @petey

**Comment by d2lam** [comment]
Changed scmUri to `bitbucket.org:username/repoId:branchName`
Build will fail until @minz1027's  change on the data-schema

File: index.js [hunk 6]
```diff
@@ -1,11 +1,83 @@
...
+        };
+
+        return new Promise((resolve, reject) =>
+            this.breaker.runCommand(options, (err, response, body) => {
+                if (err) {
+                    return reject(err);
+                }
+
+                if (response.statusCode !== 200) {
+                    return reject(new Error('Failed to contact Bitbucket API'));
```
Comment (d2lam) [inline index.js]: Added

**Comment by d2lam** [comment]
@minz1027  can you re review :D

**Pull request merged**